influxdb_store.go

```diff
@@ -119,18 +128,66 @@ func (in *InfluxDBStore) Collect(id SpanID, anns ...Annotation) error {
 		Fields:      fields,
 		Time:        time.Now().UTC(),
 	}
+	pointSizeBytes := pointMemoryUsage(reflect.ValueOf(p))
+
+	in.batchMu.Lock()
+	defer in.batchMu.Unlock()
+
+	if in.batchSizeBytes+pointSizeBytes > in.config.MaxBatchSizeBytes {
+		if in.log != nil {
+			in.log.Println("InfluxDBStore: point batch entirely dropped (this should never happen, trace data will be missing)")
+			in.log.Printf("InfluxDBStore: batchSize:%v batchSizeBytes:%v + pointSize:%v\n", len(in.batch), in.batchSizeBytes, pointSizeBytes)
+		}
+		in.batch = nil
+		in.batchSizeBytes = 0
+		return ErrQueueDropped
+	}
+	in.batchSizeBytes += pointSizeBytes
+	in.batch = append(in.batch, *p)
+	return nil
+}
+
+// flush immediately sends all pending spans in the underlying batch to
+// InfluxDB.
+func (in *InfluxDBStore) flush() error {
+	// Grab what information we need and unlock quickly to avoid contention.
+	in.batchMu.Lock()
+	batch := in.batch
+	in.batch = nil
+	in.batchSizeBytes = 0
+	in.batchMu.Unlock()
+
+	if len(batch) == 0 {
+		return nil
+	}
 
-	// A single point represents one span.
-	pts := []influxDBClient.Point{*p}
+	// Write the batch to InfluxDB.
 	bps := influxDBClient.BatchPoints{
-		Points:   pts,
+		Points:   batch,
 		Database: in.dbName,
 	}
 	_, writeErr := in.con.Write(bps)
```

Comment:
Should be safe. Their sources [don't look evil](https://github.com/influxdata/influxdb/blob/master/client/influxdb.go#L241), though their docs don't say anything about using a Client concurrently.